Split `CWallet::Create()` into `CreateNew` and `LoadExisting`

davidgumberg opened the pull request:
This PR is mostly a refactor which splits out logic used for creating wallets and for loading wallets, both of which are presently contained in `CWallet::Create()` into `CWallet::CreateNew()` and `CWallet::LoadExisting()`

The real win of this PR is that `CWallet::Create()` uses a very bad heuristic for trying to guess whether or not it is supposed to be creating a new wallet or loading an existing wallet:

https://github.com/bitcoin/bitcoin/blob/370c59261269fd9043674e0f4fd782a89e724473/src/wallet/wallet.cpp#L2882-L2885

This heuristic assumes that wallets with no `ScriptPubKeyMans` are being created, which sounds reasonable, but as demonstrated in #32112 and #32111, this can happen when the user tries to load a wallet file that is corrupted, both issues are fixed by this PR and any other misbehavior for wallet files which succeeded the broken heuristic's sniff test for new wallets.

It was already the case that every caller of `CWallet::Create()` knows whether it is creating a wallet or loading one, so we can avoid replacing this bad heuristic with another one, and just shift the burden to the caller.

davidgumberg committed:
scripted-diff: refactor: CWallet::LoadWallet->PopulateWalletFromDB

There are too many functions in CWallet with names like "Load" and
"Create", disambiguate what CWallet::LoadWallet does by renaming it to
PopulateWalletFromDB.

-BEGIN VERIFY SCRIPT-
sed -i 's|\bLoadWallet()|PopulateWalletFromDB()|g' $(git grep -l 'LoadWallet()' -- ':(exclude)src/wallet/walletdb.cpp')
-END VERIFY SCRIPT-

davidgumberg committed:
refactor: Split out wallet argument loading

This section is necessarily repetitive, makes CWallet::Create() easier
to read, and splits out functionality that will be useful when wallet
creation and loading are separated.

Review with `-color-moved=dimmed-zebra`

davidgumberg committed:
refactor: wallet: move error handling to PopulateWalletFromDB()

davidgumberg committed:
refactor: Move -walletbroadcast setting init

Modifying `fBroadcastTransactions` does not require any locks,
initialization of this wallet parameter can be relocated with all of the
other argument parsing in this function.

davidgumberg committed:
wallet: Move argument parsing to before DB load

`m_keypool_size` must be set before `CWallet::PopulateWalletFromDB()`,
in order to move parsing of `-keypool` into `CWallet::LoadWalletArgs`,
`LoadWalletArgs()` invocation in `CWallet::Create()` must be moved
before `PopulateWalletFromDB()` is called.

davidgumberg committed:
wallet: Remove redundant birth time update

Checking every SPKM in `CWallet::Create()` is not necessary, since the
only way presently for an SPKM to get added to `m_spk_managers` (the
return value of `GetAllScriptPubKeyMans()`) is through
`AddScriptPubKeyMan()`, which already invokes `MaybeUpdateBirthTime()`.

davidgumberg committed:
refactor: Wallet stats logging in its own function

This will avoid repetition when wallet creation and loading are
separated.

davidgumberg committed:
wallet: Create separate function for wallet load

Splits out logic relevant only to existing wallets in
`CWallet::Create()` into `CWallet::LoadExisting()`

davidgumberg committed:
wallet: Use CWallet::LoadExisting() for loading existing wallets.

davidgumberg committed:
test: wallet: Split create and load

davidgumberg committed:
scripted-diff: refactor: CWallet::Create() -> CreateNew()

Aside from being more legible, changing the name of `CWallet::Create()`
also validates that every instance where a new wallet is `Create()`'ed
is handled in this branch.

-BEGIN VERIFY SCRIPT-
sed -i 's|\bCreate(|CreateNew(|g' src/wallet/wallet.cpp  src/wallet/wallet.h  src/wallet/test/util.cpp src/wallet/test/wallet_tests.cpp
-END VERIFY SCRIPT-

w0xlt reviewed (commented):
Approach ACK

achow101 commented:
`PopulateWalletFromDB` inside of `CreateNew` should no longer be necessary.

It'd be nice if the load or create intention could be propagated down to the DB level as well. In `SQLiteDatabase::Open`, we pass the flags `SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE` which does both load and create. I think ideally it would be just `SQLITE_OPEN_READWRITE` for loading, and `SQLITE_OPEN_CREATE` for creating to further enforce that we expect a file to already exist for loading, and for no files to exist when creating. Similarly, we could avoid calling `TryCreateDirectories` when loading.

pablomartin4btc reviewed (commented):
Concept ACK

The current state of this code in `master` adds [confusion](https://github.com/bitcoin/bitcoin/pull/32481#discussion_r2147385870) to flows like the legacy to descriptor wallet migration that's why I think it's an important work to be done.

rkrux reviewed (commented):
Concept ACK 0f82224d3937db3f60cfaec249ac5fe3264dc3d5
I want to share my initial thoughts before I invest more time reviewing this.

---------
I am in favour of such a refactor because from a reviewing POV I lose mental cycles every time I end up going through `CWallet:Create` function that requires me to keep a mental context of whether this function was called from the creation flow or the load flow. Also, I feel this is a good time to get this done now that legacy wallets are no longer there.

However, I am slightly less inclined to review this PR in its current format because there are a few moving pieces here, and I'm afraid reviewing it thoroughly might be more time consuming and would induce less confidence in me finally a-c-king it. 

Moreover, I feel the benefits of such a refactor would be more highlighted and appreciated if this can be broken down into smaller PRs in addition to ensuring data-correctness. Couple reasons I'm suggesting a breakdown are:
1. I would like to give more attention to each of the commits by trying to understand their consequences on the wallet as a whole.
2. I don't suppose all the commits are required to be present sequentially in a single PR based on my initial look, please feel free to correct me if I'm mistaken about their ability to be present in parallel.

<details open>
<summary>PR breakdown suggestion</summary>

The first 3 can be done in parallel & the last one might be dependent on the previous ones:

1. The `PopulateWalletFromDB` related changes can be 1 PR - the first 2 commits.
3. The `fBroadcastTransactions` change is independent, can be clubbed with the birth-time-update-removal optimisation - 3rd, 6th commits.
4. The argument parsing changes also seem independent to me and can be another PR - 4th, 5th commits.
5. No strong opinion on the wallet stats logging commit, can go anywhere.
6. The last 5 commits can be 1 PR that introduces and uses the `LoadExisting` & `CreateNew` functions.
</details>


Also, I believe [this is a good suggestion](https://github.com/bitcoin/bitcoin/pull/32636#issuecomment-2920676506) that might give dividends in the future and breaking down into smaller PRs might make it easy to incorporate it.

Having suggested all this, if you think the reward for breaking down the PR is not enough because the smaller refactoring PRs might not attract reviewers or might add rebase churn, I would still review this PR later but some kind of breakdown would aid review if not the one that I suggested.

Sorry if this got too long.

pablomartin4btc commented on src/wallet/wallet.cpp:
```diff
@@ -3056,39 +2987,139 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+        LOCK(walletInstance->cs_wallet);
+
+        // ensure this wallet.dat can only be opened by clients supporting HD with chain split and expects no default key
+        walletInstance->SetMinVersion(FEATURE_LATEST);
```
Please consider #32977

pablomartin4btc commented:
Shouldn't `src/wallet/walletdb.cpp` be part of this PR? `DBErrors WalletBatch::LoadWallet(CWallet* pwallet)` is being called during wallet creation I think. If so, please also consider the **note** at #33082's description.

davidgumberg committed:
wallet: refactor: PopulateWalletFromDB use switch statement.

Co-authored-by: @w0xlt <w0xlt@users.noreply.github.com>

davidgumberg committed:
refactor: wallet: Factor out `WriteVersion()` from `PopulateWalletFromDB()`

Writing the wallet's `CLIENT_VERSION` (which indicates the last version
to have touched a wallet) needs to be done on both wallet creation and
wallet loading.

The next commit removes the `PopulateWalletFromDatabase()` call from
wallet creation, but this behavior needs to be preserved, so this commit
factors setting `CLIENT_VERSION` out of `PopulateWalletFromDatabase()`
so that wallet creation can use it in the next commit.

davidgumberg committed:
wallet: remove loading logic from CWallet::Create

davidgumberg committed:
wallet: Correctly log stats for encrypted messages.

Previously creating an encrypted wallet would result in the keypool size
incorrectly being reported as 0.

See: https://github.com/bitcoin/bitcoin/pull/32636#discussion_r2150021064

rkrux commented:
>  I would prefer to not split into multiple PR's since most of the commits here are refactors that are setting-up for splitting wallet creation and loading, the first PR(s) in the series would have very little motivation, but if you insist I am open to splitting it up.

No insistence at the moment, I will do a full review of the PR soon.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3104,6 +3104,56 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
+    }
+
+    // Load wallet
+    bool rescan_required = false;
+    auto nLoadWalletRet = walletInstance->PopulateWalletFromDB(error, warnings);
+    if (nLoadWalletRet == DBErrors::NEED_RESCAN) {
+        rescan_required = true;
+    } else if (nLoadWalletRet != DBErrors::LOAD_OK && nLoadWalletRet != DBErrors::NONCRITICAL_ERROR) {
+        return nullptr;
+    }
```
In 4c6890cdae77ca0d230227530b000f4b3c6fd77a "wallet: Create separate function for wallet load"

I think this can be simplified:

```suggestion
    auto nLoadWalletRet = walletInstance->PopulateWalletFromDB(error, warnings);
    bool rescan_required = nLoadWalletRet == DBErrors::NEED_RESCAN;
    if (nLoadWalletRet != DBErrors::LOAD_OK && nLoadWalletRet != DBErrors::NONCRITICAL_ERROR && !rescan_required) {
        return nullptr;
    }
```

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3104,6 +3104,56 @@ std::shared_ptr<CWallet> CWallet::Create(WalletContext& context, const std::stri
     return walletInstance;
 }
 
+std::shared_ptr<CWallet> CWallet::LoadExisting(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings)
```
In 4c6890cdae77ca0d230227530b000f4b3c6fd77a "wallet: Create separate function for wallet load"

When loading, `wallet_creation_flags` is unnecessary. Flags cannot be updated when loading a wallet, and having a specific warning for `WALLET_FLAG_DISABLE_PRIVATE_KEYS` seems entirely unnecessary.

davidgumberg replied:
Fixed the suggestion as I understood it, thanks for catching this.

enirox001 reviewed (commented):
Concept ACK, overall the refactor is a good attempt, and would modularize the code better. Here are review comments on a commit by commit basis. The commits are self contained and atomic, which I commend, makes review very efficient.

f92031e3b0: Good rename from LoadWallet() to PopulateWalletFromDB() with proper scripted-diff.
8b6a8a04f9: Good separation of error handling logic into PopulateWalletFromDB() with overloaded method for detailed error reporting.
f6c4454b87: Good conversion from if-else to switch statement, adds missing LOAD_FAIL case.
3a29186688: Minor optimization moving -walletbroadcast init outside locked section.
d9a3af962a: Excellent extraction of argument parsing into LoadWalletArgs() method.
347834e41b: Necessary fix moving LoadWalletArgs() before PopulateWalletFromDB() for proper init order.
39503c4667: Good cleanup removing redundant birth time update logic.
cf59f05485: Solid extraction of stats logging into LogStats() method.
29c624ad9e: Well-executed separation creating LoadExisting() method.
7f7dc32765: Completes refactoring by updating callers to use LoadExisting().
dd720fe5bd: Properly updates test infrastructure with separate create/load functions.
62e0721cae: Good extraction of version writing into WriteVersion() method for reusability. 
8ef5c41445: Successfully removes loading logic from Create(), now creation-only.
8022e04ccf: Fixes bug where encrypted wallets showed incorrect keypool size.
a6b43fa173: Good final rename from Create() to CreateNew() for clarity.

Made a few minor observations that might need attention:

davidgumberg commented on src/wallet/walletdb.h:
```diff
@@ -262,6 +262,9 @@ class WalletBatch
     DBErrors LoadWallet(CWallet* pwallet);
 
+    //! Write the given client_version.
+    bool WriteVersion(int client_version) { return m_batch->Write(DBKeys::VERSION, CLIENT_VERSION); }
```
This is out of scope, but I can address this in a follow-up.

davidgumberg commented on src/wallet/dump.cpp:
```diff
@@ -199,8 +199,7 @@ bool CreateFromDump(const ArgsManager& args, const std::string& name, const fs::
         LOCK(wallet->cs_wallet);
-        DBErrors load_wallet_ret = wallet->PopulateWalletFromDB();
-        if (load_wallet_ret != DBErrors::LOAD_OK) {
+        if (wallet->PopulateWalletFromDB() != DBErrors::LOAD_OK) {
```
This PR does not change the error handling here, but I can address this in a follow-up.

rkrux commented on src/wallet/walletdb.cpp:
```diff
@@ -1175,7 +1175,7 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
 
     if (!has_last_client || last_client != CLIENT_VERSION) // Update
-        m_batch->Write(DBKeys::VERSION, CLIENT_VERSION);
+        WriteVersion(CLIENT_VERSION);
```
In https://github.com/bitcoin/bitcoin/commit/835de2490ed925da8df77815e8e1a6457bf1dc1e "refactor: wallet: Add WriteVersion()"

Nit to differentiate this call from the other functions called above because this function is this class's member:

```diff
diff --git a/src/wallet/walletdb.cpp b/src/wallet/walletdb.cpp
index 41855b4154..387568282d 100644
--- a/src/wallet/walletdb.cpp
+++ b/src/wallet/walletdb.cpp
@@ -1175,7 +1175,7 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         return result;
 
     if (!has_last_client || last_client != CLIENT_VERSION) // Update
-        WriteVersion(CLIENT_VERSION);
+        this->WriteVersion(CLIENT_VERSION);
 
     if (any_unordered)
         result = pwallet->ReorderTransactions();
```

davidgumberg replied:
Thank you, fixed.

rkrux reviewed (commented):
Code review b29d22a3d8395865a0a79daf16d44b4eb98f571e

Looks pretty neat to me, the refactor clearly differentiates between the wallet creation & load, and makes it easy to navigate this areas of the codebase. The usage of `PopulateWalletFromDB` function in `LoadExisting` and not in `CreateNew` is also helpful in maintaining the reading flow.
+1 on also fixing the encrypted wallet logging bug.

I believe this refactor is worth checking in the master branch.

davidgumberg committed:
wallet: Delete unnecessary PopulateWalletFromDB() calls

w0xlt reviewed (commented):
ACK a3e24b0c2e2fd2d72980479ce077250430dfe1c8
The PR executes a good refactoring and changes the function names to declare their intent (create vs load) .
The identified issues (`WriteVersion` parameter, public `PopulateWalletFromDB`) are minor and appropriate for follow-up PRs.
The CI error is unrelated.

w0xlt reviewed (commented):
reACK f9de11f296df24c63d85d103c5e1e464afc9f4f4

achow101 commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -307,7 +306,7 @@ void TestLoadWallet(const std::string& name, DatabaseFormat format, std::functio
     auto database{MakeWalletDatabase(name, options, status, error)};
     auto wallet{std::make_shared<CWallet>(chain.get(), "", std::move(database))};
-    BOOST_CHECK_EQUAL(wallet->LoadWallet(), DBErrors::LOAD_OK);
+    BOOST_CHECK_EQUAL(wallet->PopulateWalletFromDB(), DBErrors::LOAD_OK);
```
In 0972785fd723b9b3c84844bf999d6e08e163ef9d "wallet: Delete unnecessary PopulateWalletFromDB() calls"

This is also unnecessary since the wallet database is once again being created in this function, so there is nothing to be loaded from it.

achow101 commented on src/wallet/test/util.cpp:
```diff
@@ -47,11 +47,35 @@ std::unique_ptr<CWallet> CreateSyncedWallet(interfaces::Chain& chain, CChain& cc
+}
+
+
 std::shared_ptr<CWallet> TestLoadWallet(std::unique_ptr<WalletDatabase> database, WalletContext& context, uint64_t create_flags)
```
In e655683ec984a73ff198ecb4a8df0d5c5b0f1f98 "test: wallet: Split create and load"

`create_flags` is now unused.

achow101 commented on src/wallet/test/util.cpp:
```diff
@@ -47,11 +47,35 @@ std::unique_ptr<CWallet> CreateSyncedWallet(interfaces::Chain& chain, CChain& cc
+
+std::shared_ptr<CWallet> TestCreateWallet(WalletContext& context)
+{
+    DatabaseOptions options;
```
In e655683ec984a73ff198ecb4a8df0d5c5b0f1f98 "test: wallet: Split create and load"

Should add a `options.require_create = true` so that `MakeWalletDatabase` can enforce that the database is being created.

davidgumberg commented on src/wallet/test/util.cpp:
Fixed.

davidgumberg commented:
([`git range-diff f9de11f...db2effa`](https://github.com/bitcoin/bitcoin/compare/f9de11f296df24c63d85d103c5e1e464afc9f4f4..db2effaca4cf82bf806596d16f9797d3692e2da7)) Pushed to address @achow101's feedback:

- Fix the bug in fallback fee behavior in https://github.com/bitcoin/bitcoin/commit/097e00f90765915b0f8353a4eb5ebb18ceae2a66 (see: https://github.com/bitcoin/bitcoin/pull/32636#discussion_r2706102550 and #34382)
- Drop `create_flags` from `TestLoadWallet()`
- Add `options.require_create = true` to `TestCreateWallet()` to enforce wallet creation.
- Add `options.require_existing = true` to `TestLoadWallet()` to enforce wallet loading.

w0xlt reviewed (commented):
reACK db2effaca4cf82bf806596d16f9797d3692e2da7

polespinasa reviewed (commented):
approach ACK db2effaca4cf82bf806596d16f9797d3692e2da7

As a contributor not really familiar with the wallet codebase this PR makes the difference when understanding the code.
Having the creating and loading logic separated makes much sense.

I reviewed all commits, they look clean and neat :) 
I just don't understand why b4a49cc7275efc16d4a4179ed34b50de5bb7367e and b15a94a618c53041e97ccfface3045a0642777e1 are on different commits, they look pretty related to me.

I don't have comments regarding the code, so far lgtm.

murchandamus reviewed (commented):
Review comments up to ae66e01"wallet: Create separate function for wallet load"

murchandamus reviewed (commented):
Thanks for the good structure of the PR that was easy to follow. I’m fairly new to looking at the wallet creation and loading logic, so I just a have a few nits that you can feel free to ignore unless you need to touch the code again. The changes look like a good improvement to me.

ACK db2effaca4cf82bf806596d16f9797d3692e2da7

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4252,7 +4252,7 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
     }
 
     // Make the local wallet
```
"make" in this context means "create the CWallet object that lives in memory", not actually making a wallet on disk.

davidgumberg commented on src/wallet/wallet.cpp:
Thanks for catching this, I've fixed this in a follow-up: #34490 (https://github.com/bitcoin/bitcoin/pull/34490/commits/dcc8bd227db84d1addfe3bebfc7897623070280f)

davidgumberg commented on src/wallet/wallet.h:
```diff
@@ -873,9 +873,12 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     static bool LoadWalletArgs(std::shared_ptr<CWallet> wallet, const WalletContext& context, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
-    /* Initializes the wallet, returns a new CWallet instance or a null pointer in case of an error */
+    /* Initializes, creates and returns a new CWallet; returns a null pointer in case of an error */
```
Responding to the review comment [below](https://github.com/bitcoin/bitcoin/pull/32636#discussion_r2229614994) here: 

What I wanted to express was that both functions, `CreateNew()` and `LoadExisting()` "initialize" a `CWallet` object, which was previously what was meant by `Create()` in `CWallet::Create()`, even though it both loaded and "created" new wallets, it always "created" a `CWallet` object.

I've improved both function descriptions in a follow-up pr #34490: https://github.com/bitcoin/bitcoin/pull/34490/commits/dcc8bd227db84d1addfe3bebfc7897623070280f to avoid making ACK's stale.

davidgumberg commented on src/wallet/wallet.h:
```diff
@@ -873,9 +873,12 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
+    /* Initializes, creates and returns a new CWallet; returns a null pointer in case of an error */
     static std::shared_ptr<CWallet> Create(WalletContext& context, const std::string& name, std::unique_ptr<WalletDatabase> database, uint64_t wallet_creation_flags, bilingual_str& error, std::vector<bilingual_str>& warnings);
 
+    /* Initializes, loads, and returns a CWallet from an existing wallet; returns a null pointer in case of an error */
```
Discussion above: https://github.com/bitcoin/bitcoin/pull/32636#discussion_r2757355261

rkrux reviewed (approved):
ACK db2effaca4cf82bf806596d16f9797d3692e2da7

This time I have gone through the commit messages in the detail, reviewed most of the commits, and compared the before & after state of the related `CWallet` code in `wallet.cpp` file to gauge the code readability.

This is a highly desirable refactor that significantly improves the wallet creation and loading code readability. It was a great pain having to read the earlier `CWallet::Create` function that handled both wallet creation & loading, along with parsing wallet arguments that ultimately made it more than 250 lines long.

This new code reads much better wherein the reader can easily differentiate between the creation and loading flow via these two fewer-than-60-lines separate functions. The separate wallet args parsing method also helps the reader greatly. The corresponding unit test methods separation is also useful.

I believe the changes of the commit 0972785fd723b9b3c84844bf999d6e08e163ef9d "wallet: Delete unnecessary PopulateWalletFromDB() calls" could have been part of the follow up PR as its presence in this one forces me to think why these calls are unnecessary, something that's not answered in the commit message.

Besides the conspicuous methods refactoring, there are also few one line code movements between the methods that logically seem alright to me and I'm also relying on the existing tests to verify their correctness. Also, it's common in refactoring PRs for the reader to mentally "skip over" the impacts of minor movements for which the success of the existing tests should supplement.

The logging bug fix of encrypted wallets creation https://github.com/bitcoin/bitcoin/pull/32636#discussion_r2150021064 in this PR is an add-on!  I also feel that after this PR is merged, the wallet encryption flow post creation can also become part of the `CWallet::CreateNew` function in the future as it involves setting up SPKMs that appears to be a part of wallet creation to me.
https://github.com/bitcoin/bitcoin/blob/eb3dbbaf30fc072fef78f57064abc744aa9c0516/src/wallet/wallet.cpp#L432-L456
